lib/Runtime/Library/JavascriptString.cpp

```diff
@@ -2228,8 +2228,55 @@ namespace Js
         }
 
         ScriptContext* scriptContext = pThis->type->GetScriptContext();
-
         ApiError error = ApiError::NoError;
+        const char16 *pThisSz = pThis->GetSz();
+        charcount_t pThisLength = pThis->GetLength();
+
+        bool isAscii = true;
+        for (charcount_t i = 0; i < pThisLength; i++)
+        {
+            if (pThisSz[i] >= 0x80)
+            {
+                isAscii = false;
+                break;
+            }
+        }
+
+        if (isAscii)
```

Comment:
@Microsoft/target-dev as FYI.
It is not safe to assume that ASCII is the same for all locales.
For example the Turkish "i" needs to handled differently than in English.
However, you might be able to do something similar to .NET Core to *check* first if you can use ASCII case mapping as a shortcut.
Take a look at the `IsAsciiCasingSameAsInvariant` in the .NET Core GitHub repo. :)

Reply:
Here is a link to the .net Core code:
https://github.com/dotnet/corefx/blob/06fb4889ad2d042b96318e765956588fa9199e46/src/Common/src/CoreLib/System/Globalization/TextInfo.cs#L587

This if statement needs to be something like this instead:
`if (IsAsciiCasingSameAsInvariant && isAscii)`

Reply:
Although, if the locale doesn't support ASCII casemapping, then you could likely bypass the string check to see if it was ASCII in the first place, as it doens't matter.

So perhaps something like this (pseudo code):
```
if (IsAsciiCasingSameAsInvariant)
{
    // check is the string is all ASCII.
    if (isAscii)
    {
        // fast case-mapping code.
        // return
    }
}

// slow case-mapping code

```

Reply:
I may just have this as `if (useInvariant)`  as @sethbrenith suggested and not worry about the case where we should follow specific rules -- as far as I understand it, `"i".toLocaleUpperCase("tr-tr")` requires special consideration but `"i".toUpperCase()` does not.

Reply:
Hmm, good point. If `toUpperCase` does not depend on any global locale state (or whatever the 'default' host locale might be), then that might be totally fine.